walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -33,6 +33,40 @@ private static boolean isAttended(Event event, MeetingRequest request) {
   }
 
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> freeTimes = new ArrayList<TimeRange>(); // Output list for possible meeting times.
+    Event[] eventsArray = events.toArray(new Event[events.size()]);
+    long duration = request.getDuration() / 30; // Meeting duration in number of half hour array slots.
+    
+    if (duration > 47) return freeTimes; // If the duration is out of bounds, we don't need to add any times.
+ 
+    boolean[] occupiedTimes = getOccupiedTimes(eventsArray, request);
+ 
+    int lowerBound = 0; // Start of Time Gap (first unoccupied slot past endPointer).
+    int upperBound = 0; // End of Time Gap (last unoccupied slot before first occupied slot after endpoint).
+ 
+    // Loop through array and find gaps between free slots.
+    // endPointer points at the first most scanned element in the array
+    // If gaps are sufficiently large, add them to list.
+    for (int endPointer = 0; endPointer < occupiedTimes.length; endPointer++) {
+      boolean slot = occupiedTimes[endPointer];
+
+      // If we find an occupied slot, set our gap upper bound to that slot.
+      if (slot == true) upperBound = endPointer;
+      else if(endPointer == occupiedTimes.length-1) upperBound = endPointer+1; // Include bound for gap size calculation.
+ 
+      if (slot == true || endPointer == occupiedTimes.length-1) {
+        int gapSize = upperBound - lowerBound; // The range between these two values is the duration of the gap
+        
+        // If the gap is sufficiently large, add a TimeRange from those bounds to our output list.
+        if (gapSize >= duration) freeTimes.add(TimeRange.fromStartEnd(getTime(lowerBound),getTime(upperBound),false));
+ 
+        // Fast forward lowerBound to next free space and update endPointer to continue from there.
+        for (lowerBound = endPointer; lowerBound < occupiedTimes.length && occupiedTimes[lowerBound] == true; lowerBound++);
```

Comment:
This line (and a couple of others here) look to me like they might be longer than 100 characters.

Reply:
Wrapped lines across the board here b9cc433